### ISSUE TITLE
serializermethodfield

### ISSUE THREAD

**Issue by whiteruby02** [wafflestudio/rookies#189]
저번 질문을 통해 userserializer를 커스터마이즈하여 4번의 그림과 같은 구조를 만드는 것까지는 인지하였습니다.
그런데 serializermethodfield는 read-only로 알고 있는데 이후에 participant나 instructor등이 수정되는 문제가 있던데,
혹시 serializer를 수정하는게 아닌, 다른 방식으로 작성해야하나요?

제가 생각하기에는 serializer를 수정하는게 다른 질문등을 보았을 때 맞는 것 같은데, serializermethodfield를 쓰면 또 read-only라 수정을 못하게 되어 혹시 제가 놓친 다른 방법이 있나 해서 여쭤봅니다ㅠ

하루정도 4번에서 serializer로 멈춰있었는데, 다시 공부하고 와도 read-only 때문에 막혔습니다ㅠ 구글링 힌트를 주셔도 감사하겠습니다.

**Comment by YeonghyeonKO** [comment]
이해하신대로 [SerializerMethodField](https://www.django-rest-framework.org/api-guide/fields/#serializermethodfield)는 read-only field로서 response body에 또 다른 serializer로 nested body를 구축하고 싶을 때 사용합니다.

맥락상 SerializerMethodField로 지정한 participant와 instructor가 represented하는데 사용될 **get_participant**, **get_instructor**(SerializerMethodFieldmethod-name argument를 default로 설정하셨다면) 함수에서 수정을 못하기 때문에 막힌 것으로 받아들였는데 맞나요?

**Comment by davin111** [comment]
‘저번 질문’에 대한 링크를 달아주시면 좋겠습니다. @whiteruby02

**Comment by whiteruby02** [comment]
@davin111  #181 입니다. 
사진은 과제 assignment.md에 4번의 
{
    "id": User id,
    "username": User username,
    "email": User email,
    "first_name": User first_name,
    "last_name": User last_name,
    "last_login": User last_login,
    "date_joined": User date_joined,
    "token": Token key (회원가입과 로그인 경우에만 포함)
    "participant": {
        "id": ParticipantProfile id,
        "university": ParticipantProfile university(string),
        "accepted": ParticipantProfile accepted(bool),
        "seminars": [
            {
                "id": Seminar id,
                "name": Seminar name(string),
                "joined_at": UserSeminar joined_at(datetime),
                "is_active": UserSeminar is_active(bool),
                "dropped_at": UserSeminar dropped_at(datetime, 정보가 없는 경우 null)
            },
            ...
        ] (참여하는 Seminar가 없는 경우 [])
    } (ParticipantProfile이 없는 경우 null),
    "instructor": {
        "id": InstructorProfile id,
        "company": InstructorProfile company(string),
        "year": InstructorProfile year(number, 정보가 없는 경우 null),
        "charge": {
            "id": Seminar id,
            "name": Seminar name(string),
            "joined_at": UserSeminar joined_at(datetime)
        } (담당하는 Seminar가 없는 경우 null)
    } (InstructorProfile이 없는 경우 null)
}
입니다.

**Comment by whiteruby02** [comment]
@YeonghyeonKO 네 맞습니다. participant와 instructor를 가져오는데 serializermethodfield를 이용해서 university, accepted, company, seminar(또 serializermethodField이용) 를 만드려고 했는데, read only 라 수정을 못해서 과제 뒷편에 participant 수정을 만족할 수 없습니다ㅠㅠ 그런데 다른 질문들을 읽어보니 다들 serializer를 통해 과제를 하신것 같아 제가 뭔가 모르는게 있나 해서 여쭤봤습니다ㅠㅠ

**Comment by davin111** [comment]
@whiteruby02 아아, 그림이라는 것이 response body를 말씀하신 것이었군요 저는 뭔가 다른 것이 있는 것인 줄 알았습니다.

사실 이건 뭔가 착각 또는 이해를 잘못 하고 계신 상황일 가능성이 높아보이는데요, `serializermethodfield는 read-only로 알고 있는데 이후에 participant나 instructor등이 수정되는 문제가 있던데,`라는 문제가 실제로 발생할 수 없어보입니다. read-only가 맞기에, 해당 `ModelSerializer`를 model을 write(create, update)할 때 사용한다면 해당 field는 아무런 영향을 끼치지 않습니다. 지금 말씀 주신 것만 가지고는 이 정도밖에 더 말씀드리기 어려울 거 같네요.

@YeonghyeonKO `SerializerMethodField`를 항상 `또 다른 serializer로 nested body를 구축하고 싶을 때 사용`하는 것은 아닙니다. 그것은 `SerializerMethodField`의 하나의 사용 방법일 뿐이고, 그냥 어떤 값을 별다른 로직 없이 `get_{field}()`에서 바로 return 해버려도 무관합니다.

**Comment by whiteruby02** [comment]
@davin111  앗 제 질문은 serializermethodfield는 read-only인데, 이후 participant와 instructor를 수정하는 스펙을 만족시키지 못하지 않나요? 라는 질문이었습니다! #181 질문에서 Userserializer를 커스터마이징 하셨다고 하셔서...
